DigitalLearningSolutions.Web/ServiceFilter/ValidateAllowedDlsSubApplication.cs

```diff
@@ -33,42 +37,83 @@ public async void OnActionExecuting(ActionExecutingContext context)
             var user = context.HttpContext.User;
             if (!user.Identity.IsAuthenticated)
             {
-                RedirectToLogin(context);
+                SetLoginRedirectResult(context);
                 return;
             }
 
-            var dlsSubApplicationParameterName = context.ActionDescriptor.Parameters?
+            if (!TryGetDlsSubApplication(context, out var application, out var applicationParameterName))
+            {
+                return;
+            }
+
+            if (await ApplicationIsInaccessibleByPage(context, application))
+            {
+                return;
+            }
+
+            ValidateUserHasPermissionForApplication(context, user, application, applicationParameterName!);
+        }
+
+        private bool TryGetDlsSubApplication(

```

Comment:
We can return a tuple of the param name and the application instead of a bool to say whether they've been successfully parsed - the values of param name and application give us more info than the bool anyway.

Reply:
Strictly speaking they don't contain the same information. It's possible for a successful parse to return a null application, and this will then pass the permissions-based validation (you can't lack permission to access a null application) but fail the whitelist-based validation (because null isn't one of the applications permitted to access the page). This is distinct from the parse failing, which will cause the validation to end after either setting or not setting the Not Found Result.

Reply:
But either way we return a NotFound right? If the application we parse out is null, I would always expect it to return a NotFound. So we don't really care if it's a successful parse in that case. Do you disagree?

Reply:
We don't necessarily return a NotFound if the parse is successful. If the endpoint has a restricted list of valid applications it will return NotFound (because 'null' is not on that list), but if it doesn't have such a list it will just wave the action through.

Reply:
If we want to return a Not Found either way I can implement that.